microovn/src/charm.py

```diff
@@ -92,49 +147,74 @@ def find_token(self, relation):
-        if self.unit.is_leader():
-            # only bootstrap if leader
-            # TODO: fix cluster bootstrapped, leader changes to new node not in
-            #       cluster, bootstrapps new cluster can be fixed via having some
-            #       other characteristic determine who manages this side of the
-            #       mirror ie: microcluster leader
-            os.system(f"microovn cluster bootstrap")
-            self._stored.in_cluster = True
-            self.unit.status = ops.ActiveStatus("cluster bootstrapped")
+    def any_token_exists(self, relation):
+        for unit in relation.units:
+            for k in relation.data[unit].keys():
+                if k.startswith(MIRROR_PREFIX) and \
+                   relation.data[unit][k] != "empty":
+                    return True
+        return False
 
     def _handle_relation_created(self, event: ops.RelationCreatedEvent):
         self.add_hostname(WORKER_RELATION)
-        self.update_mirror_state(event.relation.data)
-        if self.unit.is_leader():
-            self.update_tokens(event.relation)
+
+        token_in_cluster = False
+        if relation := self.model.get_relation(WORKER_RELATION):
+            token_in_cluster = self.any_token_exists(relation)
+
+        if not self._stored.in_cluster and self.unit.is_leader() and \
+           not token_in_cluster:
+            error, _ = call_microovn_command("cluster", "bootstrap")
+            if error:
+                logger.error("{0} unable to bootstrap cluster".format(
+                    get_hostname()))
+                self.unit.status = ops.BlockedStatus("unable to bootstrap cluster")
+                event.defer()
+                return
+            self._stored.in_cluster = True
+            self.unit.status = ops.ActiveStatus("cluster bootstrapped")
+
+        self.handle_mirror(event.relation)
 
     def _on_remove(self, event: ops.RemoveEvent):
         if self._stored.in_cluster:
-            os.system("microovn cluster remove {}".format(get_hostname()))
-
-    def _on_leader_elected(self, _: ops.RelationChangedEvent):
-        if relation := self.model.get_relation(WORKER_RELATION):
-            self.update_mirror_state(relation.data)
+            error, _ = call_microovn_command("cluster", "remove", get_hostname())
+            if error:
+                logger.error("failed removing {0} from cluster".format(
+                    get_hostname()))
 
     def _on_cluster_changed(self, event: ops.RelationChangedEvent):
         if not self._stored.in_cluster:
             if (token := self.find_token(event.relation)):
-                self.join_with_token(token)
-
-        if self.unit.is_leader() and self._stored.in_cluster:
-            return self.update_tokens(event.relation)
+                successful = self.join_with_token(token)
+                if not successful:
+                    self.unit.status = ops.BlockedStatus("Joining cluster failed")
```

Comment:
I like the way you bring fatal errors to the attention of the agent/person operating the operator!

Note that in the previous BlockedStatus call you use different capitalization for the message, consistency?